### PR TITLE
F3DEX3 cel shading and updates

### PR DESCRIPTION
**Pull request by sauraen**
Changes in this PR:

### Added cel shading for use with F3DEX3

This cel shading implementation involves drawing tris once for each "cel level", and using a F3DEX3 feature to move light level to shade alpha so it can be used in alpha compare for the thresholds. For more information, see https://www.youtube.com/watch?v=XbY6TdaeMeY .

The primary changes are:
- Material properties and GUI for cel shading
- Cel shading display list writing

### F3DEX3 updates

- Modified encodings for new geometry mode bits and some commands
- Added specular lighting and Fresnel to color
- Changed Fresnel algorithm
- Removed placeholders for some planned commands which have been scrapped

### F3DEX3 presets

Added many presets for F3DEX3, including for cel shading and other F3DEX3 features. Added a section in the readme about this.

### Other changes

- Improved warnings about invalid material settings / combinations of settings, especially w/r/t RDP silicon bugs

Like the other F3DEX3 PR (already merged), this PR does not include any changes to nodes preview in Blender, as F3DEX3 preview support will require substantial changes to this system.

One last note: This is an evolved form of the same modifications to fast64 which produced the cel shading display lists for Breath of the Wild Link and Zelda in [Triforce% OoT ACE Showcase](https://gettriforce.link). This PR is the third attempt to get those changes merged to fast64.

### CODE REVIEW

**Review by Yanis42 — COMMENTED** [review]
code looks good, I just have one small comment, I sent UI improvement ideas on discord (it's easier to talk about there)

File: fast64_internal/f3d/f3d_parser.py [hunk 1]
```diff
@@ -1579,17 +1587,35 @@ def processCommands(self, dlData, dlName, dlCommands):
                     self.setLightColor(dlData, command)
                 elif command.name[:13] == "gsSPSetLights":
                     self.setLights(dlData, command)
+                elif command.name == "gsSPAmbOcclusionAmb":
+                    mat.ao_ambient = float(int(command.params[0], 0)) / (2**16)
```
[the remark this answers is not in the document]
Comment (sauraen) [reply fast64_internal/f3d/f3d_parser.py:1591]: Unfortunately which params exist depends on which command it is, so they can't be pulled out. However I have updated it to use a utility function.

**Comment by sauraen** [comment]
Addressed @Yanis42 's comments from Discord.

**Review by Yanis42 — COMMENTED** [review]
noticed few things (and a minor UI issue I explained on Discord, among other little things) but seems ok

File: fast64_internal/oot/README.md [hunk 2]
```diff
@@ -14,17 +14,17 @@
 11. [Cutscenes](#cutscenes)
 
 ### Getting Started
-1. In the 3D view properties sidebar, go to the ``Fast64`` tab, then ``Fast64 Global Settings`` and set ``Game`` to ``OOT``.
-2. Set ``F3D Microcode`` to ``F3DEX2/LX2``.
-3. Switch to the ``OOT`` tab. In ``OOT File Settings``, set your decomp path to the path of your [OoT Decomp](https://github.com/zeldaret/oot/) repository on disk.
-4. In ``OOT Tools``, click "Add Scene" to create a basic scene.
+1. In the 3D view properties sidebar (press `n`), go to the ``Fast64`` tab, then ``Fast64 Global Settings`` and set ``Game`` to ``OOT``.
```
[the remark this answers is not in the document]
Comment (sauraen) [reply fast64_internal/oot/README.md]: Fixed

**Review by Yanis42 — APPROVED** [review]

File: fast64_internal/utility.py [hunk 3]
```diff
@@ -2,7 +2,8 @@
 from math import pi, ceil, degrees, radians, copysign
 from mathutils import *
 from .utility_anim import *
```
[the remark this answers is not in the document]
Comment (sauraen) [reply fast64_internal/utility.py:4]: Yeah, not ideal but cleaning this up is out of scope.

**Pull request merged**